shared/src/main/scala/mlscript/TypeDefs.scala

```diff
@@ -555,89 +555,92 @@ class TypeDefs extends ConstraintSolver { self: Typer =>
     var varianceUpdated: Bool = false;
 
     def fieldVarianceHelper(fieldTy: FieldType, curVariance: VarianceInfo)(implicit ctx: Ctx, tyDef: TypeDef): Unit = {
-        updateVariance(fieldTy.ub, curVariance.flip(VarianceInfo.co()))
-        fieldTy.lb.foreach(lb => updateVariance(lb, curVariance.flip(VarianceInfo.contra())))
+        fieldTy.lb.foreach(lb => updateVariance(lb, curVariance.flip))
+        updateVariance(fieldTy.ub, curVariance)
     }
 
-    def updateVariance(ty: SimpleType, curVariance: VarianceInfo)(implicit ctx: Ctx, tyDef: TypeDef): Unit = ty match {
-      case ProxyType(underlying) => updateVariance(underlying, curVariance)
-      case TraitTag(_) | ClassTag(_, _) => ()
-      case ExtrType(pol) => ()
-      case t: TypeVariable => tyDef.tvarVariance.updateWith(t) {
-        case None => Some(curVariance)
-        case Some(value) =>
-          val newVariance = value && curVariance
-          if (newVariance =/= curVariance) varianceUpdated = true;
-          Some(newVariance)
-      })
-      case RecordType(fields) => fields.foreach {
-        case (_ , fieldTy) => fieldVarianceHelper(fieldTy, curVariance)
-      }
-      case NegType(negated) =>
-        updateVariance(negated, curVariance.flip(VarianceInfo.contra()))
-      case TypeRef(defn, targs) =>
-        val typeRefDef = ctx.tyDefs.getOrElse(defn.name, throw new Exception(s"Cannot find $defn type definition in context"))
-        targs.zip(typeRefDef.tvars).foreach { case (targ, tvar) =>
-          val tvarVariance = if (curVariance === VarianceInfo.contra())
-              typeRefDef.tvarVariance.getOrElse(tvar, VarianceInfo.bi()).flip(VarianceInfo.contra())
-            else
-              typeRefDef.tvarVariance.getOrElse(tvar, VarianceInfo.bi()).flip(VarianceInfo.co())
-          updateVariance(targ, tvarVariance)
-        }
-      case ComposedType(pol, lhs, rhs) =>
-        updateVariance(lhs, curVariance)
-        updateVariance(rhs, curVariance)
-      case TypeBounds(lb, ub) =>
-        updateVariance(lb, curVariance.flip(VarianceInfo.contra()))
-        updateVariance(ub, curVariance.flip(VarianceInfo.co()))
-      case ArrayType(inner) => fieldVarianceHelper(inner, curVariance)
-      case TupleType(fields) => fields.foreach {
-          case (_ , fieldTy) => fieldVarianceHelper(fieldTy, curVariance)
+    def updateVariance(ty: SimpleType, curVariance: VarianceInfo)(implicit ctx: Ctx, tyDef: TypeDef): Unit =
+      trace(s"updateVariance(ty: $ty of ${ty.getClass()}, curVariance: $curVariance)(${tyDef.nme.name}.${tyDef.tvarVariances})") {
+        ty match {
+          case ProxyType(underlying) => updateVariance(underlying, curVariance)
+          case TraitTag(_) | ClassTag(_, _) => ()
+          case ExtrType(pol) => ()
+          case t: TypeVariable => val _ = tyDef.tvarVariances.updateWith(t) {
```

Comment:
A few points regarding the implementation. `computeVariances` maintains a mutable set of visited type variables. For any unvisited type variable it adds it to the visited set and recurses into the lb or ub types.

A question here is why are the lb and ub for `TypeVariable` expressed a `List[SimpleType]`. Why is a list needed here?

During recursion when it sees the typevariable again with it present in the visited set it updates it's variance information.

However there is one other case that needs to be handled i.e. for type variables that do not recurse. In such case their lower and upper bounds are empty and this is their first an only visit in the current branch of the type tree. So their variance information is updated in that visit itself.

Reply:
The list is not really needed but is here for historical reasons. In the future we may replace it by single bounds with intersections and unions.

> for type variables that do not recurse. In such case their lower and upper bounds are empty

No, that's not true at all. Bounds may well be non-recursive. You don't see many cases like these in simplified `Type`s (though this will change soon), but the internal representations often have non-recursive bounds.